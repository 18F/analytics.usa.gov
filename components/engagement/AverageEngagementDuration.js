import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../js/lib/renderblock";
import formatters from "../../js/lib/formatters";

function AverageEngagementDuration({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/engagement-duration-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    // The average engagement duration block is the engagement duration in
    // seconds divided by the number of sessions. The result is rounded and
    // formatted to a readable amount of time.
    const initEngagementDurationChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const metrics = data.data[0];
            const avgEngagementDuration =
              parseInt(metrics.userEngagementDuration) /
              parseInt(metrics.visits);
            selection.text(
              formatters.secondsToReadableTime(avgEngagementDuration),
            );
          }),
        );
      return result;
    };
    initEngagementDurationChart().catch(console.error);
  }, []);

  return (
    <>
      <div ref={ref}>
        <div className="data"></div>
      </div>
    </>
  );
}

AverageEngagementDuration.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default AverageEngagementDuration;
