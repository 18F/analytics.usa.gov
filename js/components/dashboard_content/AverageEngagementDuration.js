import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import formatters from "../../lib/chart_helpers/formatters";
import Tooltip from "../tooltip/Tooltip";

/**
 * Retrieves the engagement duration report from the passed data URL and creates
 * a visualization for the average engagement duration of users for the
 * current agency. The average duration is calculated by taking the total user
 * engagement duration and dividing by the number of visits.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function AverageEngagementDuration({ dataHrefBase }) {
  const jsonDataURL = `${dataHrefBase}/engagement-duration-30-days.json`;
  const ref = useRef(null);
  const [engagementDurationData, setEngagementDurationData] = useState(null);

  useEffect(() => {
    const initEngagementDurationsChart = async () => {
      if (!engagementDurationData) {
        const data = await DataLoader.loadDailyReportJSON(jsonDataURL);
        await setEngagementDurationData(data);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder
          .setElement(ref.current)
          .setData(engagementDurationData)
          .setRenderer((selection, data) => {
            const metrics = data.data[0] || {};
            const avgEngagementDuration =
              parseInt(metrics.userEngagementDuration) /
              parseInt(metrics.visits);
            selection.text(
              formatters.secondsToReadableTime(avgEngagementDuration),
            );
          })
          .build();
      }
    };
    initEngagementDurationsChart().catch(console.error);
  }, [engagementDurationData]);

  return (
    <>
      <a
        className="chart__title"
        href="/definitions#report_historical_average_engagement_time"
      >
        <Tooltip
          position="top"
          content="The average length of time in which users were engaged with the site or application per session."
        >
          Average Engagement Time Per Session
        </Tooltip>
      </a>
      <div ref={ref}>
        <div className="data chart__rate padding-top-05"></div>
      </div>
    </>
  );
}

AverageEngagementDuration.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default AverageEngagementDuration;
