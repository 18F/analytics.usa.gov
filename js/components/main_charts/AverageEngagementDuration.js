import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import formatters from "../../lib/chart_helpers/formatters";

/**
 * Retrieves the engagement duration report from the passed data URL and creates
 * a visualization for the average engagement duration of users for the
 * current agency. The average duration is calculated by taking the total user
 * engagement duration and dividing by the number of visits.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @returns {import('react').ReactElement} The rendered element
 */
function AverageEngagementDuration({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/engagement-duration-30-days.json`;
  const ref = useRef(null);

  useEffect(() => {
    /**
     * The average engagement duration block is the engagement duration in
     * seconds divided by the number of sessions. The result is rounded and
     * formatted to a readable amount of time.
     *
     * @returns {Promise} resolves when render is complete
     */
    const initEngagementDurationChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.loadAndRender().render((selection, data) => {
            const metrics = data.data[0];
            const avgEngagementDuration =
              parseInt(metrics.userEngagementDuration) /
              parseInt(metrics.visits);
            selection.text(
              formatters.secondsToReadableTime(avgEngagementDuration),
            );
          }),
        );
      return result;
    };
    initEngagementDurationChart().catch(console.error);
  }, []);

  return (
    <>
      <div ref={ref}>
        <div className="data"></div>
      </div>
    </>
  );
}

AverageEngagementDuration.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default AverageEngagementDuration;
