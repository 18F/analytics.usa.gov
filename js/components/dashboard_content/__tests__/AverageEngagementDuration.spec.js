import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import AverageEngagementDuration from "../AverageEngagementDuration";
import DataLoader from "../../../lib/data_loader";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadDailyReportJSON: jest.fn(),
}));

describe("AverageEngagementDuration", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <AverageEngagementDuration dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "user-engagement-duration",
        query: {
          metrics: [
            {
              name: "userEngagementDuration",
            },
            {
              name: "sessions",
            },
          ],
          dateRanges: [
            {
              startDate: "30daysAgo",
              endDate: "yesterday",
            },
          ],
          limit: "10000",
          property: "properties/397708109",
        },
        meta: {
          name: "User engagement duration (30 Days)",
          description:
            "Top 20 sources that initiated a session for the last 30 days, measured by visits, for all sites.",
        },
        data: [
          {
            userEngagementDuration: "101105154773",
            visits: "1616841012",
          },
        ],
        totals: {
          visits: 1616841012,
        },
        taken_at: "2024-03-19T20:58:28.812Z",
      };

      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <AverageEngagementDuration dataHrefBase="http://www.example.com/data/" />,
      );
      await waitFor(() => screen.getByText("1 min 3 sec"));
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadDailyReportJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <AverageEngagementDuration dataHrefBase="http://www.example.com/data/" />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
