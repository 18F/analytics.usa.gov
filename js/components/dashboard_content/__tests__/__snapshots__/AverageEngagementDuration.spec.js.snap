// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AverageEngagementDuration when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    >
      1 min 3 sec
    </div>
  </div>
</DocumentFragment>
`;

exports[`AverageEngagementDuration when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    />
  </div>
</DocumentFragment>
`;

exports[`AverageEngagementDuration when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <div>
    <div
      class="data"
    />
  </div>
</DocumentFragment>
`;
